storage/s3.go

```diff
@@ -658,9 +679,27 @@ func newSession(opts S3Options) (*session.Session, error) {
 		sess.Config.Region = aws.String(endpoints.UsEast1RegionID)
 	}
 
+	// get region of the bucket and create session accordingly
+	if opts.bucket != "" {
+		region, err := s3manager.GetBucketRegion(context.Background(), sess, string(opts.bucket), "")
```

Comment:
Since creating sessions needs a network call after this update, we should use the main context instead of `context.Background()` (requests should be cancellable). You can update NewClient function to take context from parameters.   

```
func NewClient(ctx context.Context, url *url.URL) (Storage, error)
 ```